File: ats/src/main/java/fr/automated/trading/systems/portfoliosmanager/Trade.java
package fr.automated.trading.systems.portfoliosmanager;

import fr.automated.trading.systems.utils.utils.AtsLogger;

public class Trade {
	
	private String ref;
	private double priceIn;
	private double priceOut;
	private double pricePredicted = 0;
	private int volume = 1;
	private PricesConstants position;
	private boolean isClosed = false;

	public Trade(String ref, double priceIn, double pricePredicted, int volume, PricesConstants position) {
		setRef(ref);
		setPriceIn(priceIn);
		setPricePredicted(pricePredicted);
		setVolume(volume);
        setPosition(position);
		setClosed(false);
	}
	
	public void setPosition(PricesConstants position) {
		this.position = position;
	}
	
	public PricesConstants getPosition() {
		return position;
	}

	public String getRef() {
		return ref;
	}

	public void setRef(String ref) {
		this.ref = ref;
	}

	public double getPriceIn() {
		return priceIn;
	}

	public void setPriceIn(double priceIn) {
		this.priceIn = priceIn;
	}

	public double getPriceOut() {
		return priceOut;
	}

	public void setPriceOut(double priceOut) {
		this.priceOut = priceOut;
		setClosed(true);
	}

	public int getVolume() {
		return volume;
	}

	public void setVolume(int volume) {
		this.volume = volume;
	}
	
	public void setPricePredicted(double pricePredicted) {
		this.pricePredicted = pricePredicted;
	}
	
	public void debug() {
		AtsLogger.log("Trade " + ref + " position = " + getPosition() + " : priceIn = " + priceIn + " pricePredicted = " + pricePredicted + " priceOut = " + priceOut + " volume = " + volume);
	}

	public boolean isClosed() {
		return isClosed;
	}

	public void setClosed(boolean isClosed) {
		this.isClosed = isClosed;
	}
}

File: cpp/src/Trade.cpp
#include "../include/Trade.h"
#include <iostream>
#include <stdlib.h>

using namespace std;

Trade::Trade(string ref, double priceIn, double pricePredicted, unsigned int volume, string position)
: _ref(ref), _priceIn(priceIn), _pricePredicted(pricePredicted), _volume(volume), _position(position)
{
    setClosed(false);
}

Trade::~Trade()
{

}

void Trade::setPosition(string position)
{
    if(position == "BUY" || position == "SELL")
        _position = position;
    else
    {
        cout << "Invalid position" << endl;
        exit(0);
    }
}

string Trade::getPosition()
{
    return _position;
}

string Trade::getRef()
{
    return _ref;
}

void Trade::setRef(string ref)
{
    _ref = ref;
}

double Trade::getPriceIn()
{
    return _priceIn;
}

void Trade::setPriceIn(double priceIn)
{
    _priceIn = priceIn;
}

double Trade::getPriceOut()
{
    return _priceOut;
}

void Trade::setPriceOut(double priceOut)
{
    _priceOut = priceOut;
}

unsigned int Trade::getVolume()
{
    return _volume;
}

void Trade::setVolume(unsigned int volume)
{
    _volume = volume;
}

void Trade::setPricePredicted(double pricePredicted)
{
    _pricePredicted = pricePredicted;
}

void Trade::debug()
{
    cout << "Trade " << _ref << " position : " << _position;
    cout << " priceIn : " << _priceIn;
    cout << " pricePredicted : " << _pricePredicted;
    cout << " priceOut : " << _priceOut;
    cout << " volume : " << _volume << endl;
}

bool Trade::isClosed()
{
    return _isClosed;
}

void Trade::setClosed(bool isClosed)
{
    _isClosed = isClosed;
}
